core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java

```diff
@@ -114,119 +119,131 @@ public AggregationMetadata specialize(BoundSignature boundSignature)
     }
 
-    public static void inputShortDecimal(LongDecimalWithOverflowAndLongState state, Block block, int position)
+    public static void inputShortDecimal(Int128State decimalState, LongState counterState, NullableLongState overflowState, Block block, int position)
     {
-        state.addLong(1); // row counter
+        long[] decimal = decimalState.getArray();
+        int decimalOffset = decimalState.getArrayOffset();
 
-        state.setNotNull();
-
-        long[] decimal = state.getDecimalArray();
-        int offset = state.getDecimalArrayOffset();
+        decimalState.setNotNull(true);
+        counterState.setValue(counterState.getValue() + 1);
 
         long rightLow = block.getLong(position, 0);
         long rightHigh = rightLow >> 63;
 
         long overflow = addWithOverflow(
-                decimal[offset],
-                decimal[offset + 1],
+                decimal[decimalOffset],
+                decimal[decimalOffset + 1],
                 rightHigh,
                 rightLow,
                 decimal,
-                offset);
+                decimalOffset);
 
-        state.addOverflow(overflow);
+        overflow += overflowState.getValue();
+        overflowState.setNull(overflow == 0 & overflowState.isNull());
+        overflowState.setValue(overflow);
     }
 
-    public static void inputLongDecimal(LongDecimalWithOverflowAndLongState state, Block block, int position)
+    public static void inputLongDecimal(Int128State decimalState, LongState counterState, NullableLongState overflowState, Block block, int position)
     {
-        state.addLong(1); // row counter
-
-        state.setNotNull();
+        long[] decimal = decimalState.getArray();
+        int decimalOffset = decimalState.getArrayOffset();
 
-        long[] decimal = state.getDecimalArray();
-        int offset = state.getDecimalArrayOffset();
+        counterState.setValue(counterState.getValue() + 1);
+        decimalState.setNotNull(true);
 
         long rightHigh = block.getLong(position, 0);
         long rightLow = block.getLong(position, SIZE_OF_LONG);
 
         long overflow = addWithOverflow(
-                decimal[offset],
-                decimal[offset + 1],
+                decimal[decimalOffset],
+                decimal[decimalOffset + 1],
                 rightHigh,
                 rightLow,
                 decimal,
-                offset);
+                decimalOffset);
 
-        state.addOverflow(overflow);
+        overflow += overflowState.getValue();
+        overflowState.setNull(overflow == 0 & overflowState.isNull());
+        overflowState.setValue(overflow);
     }
 
-    public static void combine(LongDecimalWithOverflowAndLongState state, LongDecimalWithOverflowAndLongState otherState)
+    public static void combine(Int128State decimalState, LongState counterState, NullableLongState overflowState, Int128State otherDecimalState, LongState otherCounterState, NullableLongState otherOverflowState)
```

Comment:
you should use annotated functions like in `io.trino.operator.aggregation.AverageAggregations`

Reply:
you should base your code of these two commits:
https://github.com/dain/trino/commit/77365dbb5bfb14c69fb14d82bcf15001067dcfae
https://github.com/dain/trino/commit/a86e3840cc3e046dd764fa56a06e128461d2d78a

Reply:
We want annotated aggregations (I missed that during initial code skimming). I don't think https://github.com/trinodb/trino/pull/11477 should be blocked on this PR as this PR requires investigation